File: sknetwork/visualization/dendrograms.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created in April 2020
@author: Thomas Bonald <bonald@enst.fr>
"""
from typing import Iterable, Optional

import numpy as np

from sknetwork.hierarchy.postprocess import cut_straight
from sknetwork.visualization.colors import STANDARD_COLORS


def get_index(dendrogram, reorder=True):
    """Index nodes for pretty dendrogram."""
    n = dendrogram.shape[0] + 1
    tree = {i: [i] for i in range(n)}
    for t in range(n - 1):
        i = int(dendrogram[t, 0])
        j = int(dendrogram[t, 1])
        left: list = tree.pop(i)
        right: list = tree.pop(j)
        if reorder and len(left) < len(right):
            tree[n + t] = right + left
        else:
            tree[n + t] = left + right
    return list(tree.values())[0]


def svg_dendrogram_top(dendrogram, names, width, height, margin, margin_text, scale, line_width, n_clusters,
                       color, colors, font_size, reorder, rotate_names):
    """Dendrogram as SVG image with root on top."""

    # scaling
    height *= scale
    width *= scale

    # positioning
    labels = cut_straight(dendrogram, n_clusters, return_dendrogram=False)
    index = get_index(dendrogram, reorder)
    n = len(index)
    unit_height = height / dendrogram[-1, 2]
    unit_width = width / n
    height_basis = margin + height
    position = {index[i]: (margin + i * unit_width, height_basis) for i in range(n)}
    label = {i: l for i, l in enumerate(labels)}
    width += 2 * margin
    height += 2 * margin
    if names is not None:
        text_length = np.max(np.array([len(str(name)) for name in names]))
        height += text_length * font_size * .5 + margin_text

    svg = """<svg width="{}" height="{}"  xmlns="http://www.w3.org/2000/svg">""".format(width, height)

    # text
    if names is not None:
        for i in range(n):
            x, y = position[i]
            x -= margin_text
            y += margin_text
            text = str(names[i]).replace('&', ' ')
            if rotate_names:
                svg += """<text x="{}" y="{}"  transform="rotate(60, {}, {})" font-size="{}">{}</text>""" \
                    .format(x, y, x, y, font_size, text)
            else:
                y += margin_text
                svg += """<text x="{}" y="{}"  font-size="{}">{}</text>""" \
                    .format(x, y, font_size, text)

    # tree
    for t in range(n - 1):
        i = int(dendrogram[t, 0])
        j = int(dendrogram[t, 1])
        x1, y1 = position.pop(i)
        x2, y2 = position.pop(j)
        l1 = label.pop(i)
        l2 = label.pop(j)
        if l1 == l2:
            line_color = colors[l1 % len(colors)]
        else:
            line_color = color
        x = .5 * (x1 + x2)
        y = height_basis - dendrogram[t, 2] * unit_height
        svg += """<path stroke-width="{}" stroke="{}" d="M {} {} {} {}" />"""\
            .format(line_width, line_color, x1, y1, x1, y)
        svg += """<path stroke-width="{}" stroke="{}" d="M {} {} {} {}" />"""\
            .format(line_width, line_color, x2, y2, x2, y)
        svg += """<path stroke-width="{}" stroke="{}" d="M {} {} {} {}" />"""\
            .format(line_width, line_color, x1, y, x2, y)
        position[n + t] = (x, y)
        label[n + t] = l1

    svg += '</svg>'
    return svg


def svg_dendrogram_left(dendrogram, names, width, height, margin, margin_text, scale, line_width, n_clusters,
                        color, colors, font_size, reorder):
    """Dendrogram as SVG image with root on left side."""

    # scaling
    height *= scale
    width *= scale

    # positioning
    labels = cut_straight(dendrogram, n_clusters, return_dendrogram=False)
    index = get_index(dendrogram, reorder)
    n = len(index)
    unit_height = height / n
    unit_width = width / dendrogram[-1, 2]
    width_basis = width + margin
    position = {index[i]: (width_basis, margin + i * unit_height) for i in range(n)}
    label = {i: l for i, l in enumerate(labels)}
    width += 2 * margin
    height += 2 * margin
    if names is not None:
        text_length = np.max(np.array([len(str(name)) for name in names]))
        width += text_length * font_size * .5 + margin_text

    svg = """<svg width="{}" height="{}"  xmlns="http://www.w3.org/2000/svg">""".format(width, height)

    # text
    if names is not None:
        for i in range(n):
            x, y = position[i]
            x += margin_text
            y += unit_height / 3
            text = str(names[i]).replace('&', ' ')
            svg += """<text x="{}" y="{}" font-size="{}">{}</text>""" \
                .format(x, y, font_size, text)

    # tree
    for t in range(n - 1):
        i = int(dendrogram[t, 0])
        j = int(dendrogram[t, 1])
        x1, y1 = position.pop(i)
        x2, y2 = position.pop(j)
        l1 = label.pop(i)
        l2 = label.pop(j)
        if l1 == l2:
            line_color = colors[l1 % len(colors)]
        else:
            line_color = color
        y = .5 * (y1 + y2)
        x = width_basis - dendrogram[t, 2] * unit_width
        svg += """<path stroke-width="{}" stroke="{}" d="M {} {} {} {}" />"""\
            .format(line_width, line_color, x1, y1, x, y1)
        svg += """<path stroke-width="{}" stroke="{}" d="M {} {} {} {}" />"""\
            .format(line_width, line_color, x2, y2, x, y2)
        svg += """<path stroke-width="{}" stroke="{}" d="M {} {} {} {}" />"""\
            .format(line_width, line_color, x, y1, x, y2)
        position[n + t] = (x, y)
        label[n + t] = l1

    svg += '</svg>'

    return svg


def visualize_dendrogram(dendrogram: np.ndarray, names: Optional[np.ndarray] = None, rotate: bool = False,
                         width: float = 400, height: float = 300, margin: float = 10, margin_text: float = 5,
                         scale: float = 1, line_width: float = 2, n_clusters: int = 2, color: str = 'black',
                         colors: Optional[Iterable] = None, font_size: int = 12, reorder: bool = False,
                         rotate_names: bool = True, filename: Optional[str] = None):
    """Return the image of a dendrogram in SVG format.

    Parameters
    ----------
    dendrogram :
        Dendrogram to display.
    names :
        Names of leaves.
    rotate :
        If ``True``, rotate the tree so that the root is on the left.
    width :
        Width of the image (margins excluded).
    height :
        Height of the image (margins excluded).
    margin :
        Margin.
    margin_text :
        Margin between leaves and their names, if any.
    scale :
        Scaling factor.
    line_width :
        Line width.
    n_clusters :
        Number of coloured clusters to display.
    color :
        Default SVG color for the dendrogram.
    colors :
        SVG colors of the clusters of the dendrogram (optional).
    font_size :
        Font size.
    reorder :
        If ``True``, reorder leaves so that left subtree has more leaves than right subtree.
    rotate_names :
        If ``True``, rotate names of leaves (only valid if **rotate** is ``False``).
    filename :
        Filename for saving image (optional).

    Example
    -------
    >>> dendrogram = np.array([[0, 1, 1, 2], [2, 3, 2, 3]])
    >>> from sknetwork.visualization import svg_dendrogram
    >>> image = svg_dendrogram(dendrogram)
    >>> image[1:4]
    'svg'
    """
    if colors is None:
        colors = STANDARD_COLORS
    elif isinstance(colors, dict):
        colors = np.array(list(colors.values()))
    elif isinstance(colors, list):
        colors = np.array(colors)

    if rotate:
        svg = svg_dendrogram_left(dendrogram, names, width, height, margin, margin_text, scale, line_width, n_clusters,
                                  color, colors, font_size, reorder)
    else:
        svg = svg_dendrogram_top(dendrogram, names, width, height, margin, margin_text, scale, line_width, n_clusters,
                                 color, colors, font_size, reorder, rotate_names)

    if filename is not None:
        with open(filename + '.svg', 'w') as f:
            f.write(svg)

    return svg


def svg_dendrogram(dendrogram: np.ndarray, names: Optional[np.ndarray] = None, rotate: bool = False, width: float = 400,
                   height: float = 300, margin: float = 10, margin_text: float = 5, scale: float = 1,
                   line_width: float = 2, n_clusters: int = 2, color: str = 'black', colors: Optional[Iterable] = None,
                   font_size: int = 12, reorder: bool = False, rotate_names: bool = True,
                   filename: Optional[str] = None):
    """Return the image of a dendrogram in SVG format.

    Alias for visualize_dendrogram.

    Parameters
    ----------
    dendrogram :
        Dendrogram to display.
    names :
        Names of leaves.
    rotate :
        If ``True``, rotate the tree so that the root is on the left.
    width :
        Width of the image (margins excluded).
    height :
        Height of the image (margins excluded).
    margin :
        Margin.
    margin_text :
        Margin between leaves and their names, if any.
    scale :
        Scaling factor.
    line_width :
        Line width.
    n_clusters :
        Number of coloured clusters to display.
    color :
        Default SVG color for the dendrogram.
    colors :
        SVG colors of the clusters of the dendrogram (optional).
    font_size :
        Font size.
    reorder :
        If ``True``, reorder leaves so that left subtree has more leaves than right subtree.
    rotate_names :
        If ``True``, rotate names of leaves (only valid if **rotate** is ``False``).
    filename :
        Filename for saving image (optional).
    """
    return visualize_dendrogram(dendrogram, names, rotate, width, height, margin, margin_text, scale, line_width,
                                n_clusters, color, colors, font_size, reorder, rotate_names, filename)

File: sknetwork-rs/src/visualization/dendrograms.rs
use crate::hierarchy::postprocess::Dendrogram;
use crate::hierarchy::postprocess::cut_straight;
use crate::visualization::colors::STANDARD_COLORS;
use std::fmt::Write as FmtWrite;

fn escape_xml_text(input: &str) -> String {
    let mut out = String::with_capacity(input.len());
    for ch in input.chars() {
        match ch {
            '&' => out.push_str("&amp;"),
            '<' => out.push_str("&lt;"),
            '>' => out.push_str("&gt;"),
            '"' => out.push_str("&quot;"),
            '\'' => out.push_str("&apos;"),
            _ => out.push(ch),
        }
    }
    out
}

fn get_index(dendrogram: &Dendrogram, reorder: bool) -> Vec<usize> {
    let n = dendrogram.len() + 1;
    let mut tree: std::collections::HashMap<usize, Vec<usize>> = (0..n).map(|i| (i, vec![i])).collect();
    for (t, row) in dendrogram.iter().enumerate().take(n - 1) {
        let i = row[0] as usize;
        let j = row[1] as usize;
        let left = tree.remove(&i).unwrap_or_default();
        let right = tree.remove(&j).unwrap_or_default();
        let merged = if reorder && left.len() < right.len() {
            right.into_iter().chain(left.into_iter()).collect()
        } else {
            left.into_iter().chain(right.into_iter()).collect()
        };
        tree.insert(n + t, merged);
    }
    tree.into_values().next().unwrap_or_default()
}

/// Computes svg dendrogram top.
pub fn svg_dendrogram_top(
    dendrogram: &Dendrogram,
    names: Option<&[String]>,
    width: f64,
    height: f64,
    n_clusters: usize,
    reorder: bool,
) -> String {
    if dendrogram.is_empty() {
        return "<svg></svg>".to_string();
    }
    let labels = cut_straight(dendrogram, Some(n_clusters), None, true).unwrap_or_default();
    let colors: Vec<String> = STANDARD_COLORS.iter().map(|s| s.to_string()).collect();
    let index = get_index(dendrogram, reorder);
    let n = index.len();
    let h_max = dendrogram.last().map(|r| r[2]).unwrap_or(1.0).max(1.0);
    let unit_height = height / h_max;
    let unit_width = width / (n as f64).max(1.0);
    let mut pos: std::collections::HashMap<usize, (f64, f64)> = (0..n)
        .map(|k| (index[k], ((k as f64) * unit_width + 10.0, height + 10.0)))
        .collect();
    let mut label_map: std::collections::HashMap<usize, usize> =
        (0..labels.len()).map(|i| (i, labels[i])).collect();
    let mut svg = String::with_capacity(256 + n * 96);
    let _ = write!(
        svg,
        "<svg width=\"{}\" height=\"{}\" xmlns=\"http://www.w3.org/2000/svg\">",
        width + 20.0,
        height + 40.0
    );
    for (t, row) in dendrogram.iter().enumerate().take(n - 1) {
        let i = row[0] as usize;
        let j = row[1] as usize;
        let (x1, y1) = pos.remove(&i).unwrap_or((0.0, 0.0));
        let (x2, y2) = pos.remove(&j).unwrap_or((0.0, 0.0));
        let l1 = label_map.remove(&i).unwrap_or(0);
        let l2 = label_map.remove(&j).unwrap_or(0);
        let color = if l1 == l2 {
            &colors[l1 % colors.len()]
        } else {
            "black"
        };
        let x = 0.5 * (x1 + x2);
        let y = height + 10.0 - row[2] * unit_height;
        svg.push_str(&format!("<path stroke-width=\"1\" stroke=\"{}\" d=\"M {} {} {} {}\" />", color, x1 as i32, y1 as i32, x1 as i32, y as i32));
        svg.push_str(&format!("<path stroke-width=\"1\" stroke=\"{}\" d=\"M {} {} {} {}\" />", color, x2 as i32, y2 as i32, x2 as i32, y as i32));
        svg.push_str(&format!("<path stroke-width=\"1\" stroke=\"{}\" d=\"M {} {} {} {}\" />", color, x1 as i32, y as i32, x2 as i32, y as i32));
        pos.insert(n + t, (x, y));
        label_map.insert(n + t, l1);
    }
    if let Some(node_names) = names {
        for i in 0..n.min(node_names.len()) {
            let _ = write!(
                svg,
                "<text x=\"{}\" y=\"{}\" font-size=\"12\">{}</text>",
                (i as f64 * unit_width + 5.0) as i32,
                (height + 25.0) as i32,
                escape_xml_text(&node_names[i])
            );
        }
    }
    svg.push_str("</svg>");
    svg
}

/// Computes svg dendrogram left.
pub fn svg_dendrogram_left(
    dendrogram: &Dendrogram,
    names: Option<&[String]>,
    width: f64,
    height: f64,
    n_clusters: usize,
    reorder: bool,
) -> String {
    if dendrogram.is_empty() {
        return "<svg></svg>".to_string();
    }
    let labels = cut_straight(dendrogram, Some(n_clusters), None, true).unwrap_or_default();
    let colors: Vec<String> = STANDARD_COLORS.iter().map(|s| s.to_string()).collect();
    let index = get_index(dendrogram, reorder);
    let n = index.len();
    let h_max = dendrogram.last().map(|r| r[2]).unwrap_or(1.0).max(1.0);
    let unit_height = height / (n as f64).max(1.0);
    let unit_width = width / h_max;
    let mut pos: std::collections::HashMap<usize, (f64, f64)> = (0..n)
        .map(|k| (index[k], (width + 10.0, (k as f64) * unit_height + 10.0)))
        .collect();
    let mut label_map: std::collections::HashMap<usize, usize> =
        (0..labels.len()).map(|i| (i, labels[i])).collect();
    let mut svg = String::with_capacity(256 + n * 96);
    let _ = write!(
        svg,
        "<svg width=\"{}\" height=\"{}\" xmlns=\"http://www.w3.org/2000/svg\">",
        width + 60.0,
        height + 20.0
    );
    for (t, row) in dendrogram.iter().enumerate().take(n - 1) {
        let i = row[0] as usize;
        let j = row[1] as usize;
        let (x1, y1) = pos.remove(&i).unwrap_or((0.0, 0.0));
        let (x2, y2) = pos.remove(&j).unwrap_or((0.0, 0.0));
        let l1 = label_map.remove(&i).unwrap_or(0);
        let l2 = label_map.remove(&j).unwrap_or(0);
        let color = if l1 == l2 {
            &colors[l1 % colors.len()]
        } else {
            "black"
        };
        let y = 0.5 * (y1 + y2);
        let x = width + 10.0 - row[2] * unit_width;
        svg.push_str(&format!("<path stroke-width=\"1\" stroke=\"{}\" d=\"M {} {} {} {}\" />", color, x1 as i32, y1 as i32, x as i32, y1 as i32));
        svg.push_str(&format!("<path stroke-width=\"1\" stroke=\"{}\" d=\"M {} {} {} {}\" />", color, x2 as i32, y2 as i32, x as i32, y2 as i32));
        svg.push_str(&format!("<path stroke-width=\"1\" stroke=\"{}\" d=\"M {} {} {} {}\" />", color, x as i32, y1 as i32, x as i32, y2 as i32));
        pos.insert(n + t, (x, y));
        label_map.insert(n + t, l1);
    }
    if let Some(node_names) = names {
        for i in 0..n.min(node_names.len()) {
            let _ = write!(
                svg,
                "<text x=\"{}\" y=\"{}\" font-size=\"12\">{}</text>",
                (width + 15.0) as i32,
                ((i as f64) * unit_height + 14.0) as i32,
                escape_xml_text(&node_names[i])
            );
        }
    }
    svg.push_str("</svg>");
    svg
}

/// Computes visualize dendrogram.
pub fn visualize_dendrogram(
    dendrogram: &Dendrogram,
    names: Option<&[String]>,
    rotate: bool,
    width: f64,
    height: f64,
    n_clusters: usize,
    reorder: bool,
) -> String {
    if rotate {
        svg_dendrogram_left(dendrogram, names, width, height, n_clusters, reorder)
    } else {
        svg_dendrogram_top(dendrogram, names, width, height, n_clusters, reorder)
    }
}

/// Computes svg dendrogram.
pub fn svg_dendrogram(
    dendrogram: &Dendrogram,
    names: Option<&[String]>,
    rotate: bool,
    width: f64,
    height: f64,
    n_clusters: usize,
    reorder: bool,
) -> String {
    visualize_dendrogram(dendrogram, names, rotate, width, height, n_clusters, reorder)
}

#[derive(Debug, Clone)]
/// DendrogramVizOptions value.
pub struct DendrogramVizOptions {
    /// Rotate value.
    pub rotate: bool,
    /// Width value.
    pub width: f64,
    /// Height value.
    pub height: f64,
    /// N Clusters value.
    pub n_clusters: usize,
    /// Reorder value.
    pub reorder: bool,
}

impl Default for DendrogramVizOptions {
    fn default() -> Self {
        Self {
            rotate: false,
            width: 400.0,
            height: 300.0,
            n_clusters: 2,
            reorder: false,
        }
    }
}

/// Computes visualize dendrogram with options.
pub fn visualize_dendrogram_with_options(
    dendrogram: &Dendrogram,
    options: Option<DendrogramVizOptions>,
) -> String {
    let opts = options.unwrap_or_default();
    visualize_dendrogram(
        dendrogram,
        None,
        opts.rotate,
        opts.width,
        opts.height,
        opts.n_clusters,
        opts.reorder,
    )
}

/// Computes svg dendrogram with options.
pub fn svg_dendrogram_with_options(
    dendrogram: &Dendrogram,
    options: Option<DendrogramVizOptions>,
) -> String {
    visualize_dendrogram_with_options(dendrogram, options)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_visualize_dendrogram() {
        let d = vec![
            [0.0, 1.0, 1.0, 2.0],
            [2.0, 3.0, 2.0, 3.0],
        ];
        let image = visualize_dendrogram(&d, None, false, 400.0, 300.0, 2, false);
        assert_eq!(&image[1..4], "svg");
        let image = svg_dendrogram(&d, None, true, 400.0, 300.0, 2, true);
        assert_eq!(&image[1..4], "svg");
        let image = svg_dendrogram_top(&d, None, 400.0, 300.0, 2, false);
        assert_eq!(&image[1..4], "svg");
        let image = svg_dendrogram_left(&d, None, 400.0, 300.0, 2, false);
        assert_eq!(&image[1..4], "svg");
        let image = visualize_dendrogram_with_options(&d, None);
        assert_eq!(&image[1..4], "svg");
        let image = svg_dendrogram_with_options(&d, None);
        assert_eq!(&image[1..4], "svg");
        let names = vec!["<a&\"b'>".to_string(), "n2".to_string(), "n3".to_string()];
        let image = svg_dendrogram_top(&d, Some(&names), 400.0, 300.0, 2, false);
        assert!(image.contains("&lt;a&amp;&quot;b&apos;&gt;"));
    }
}
